src/middlewared/middlewared/plugins/virt/global.py

```diff
@@ -0,0 +1,372 @@
+            if i['type'] == 'BRIDGE'
+        })
+        return choices
+
+    @api_method(VirtGlobalPoolChoicesArgs, VirtGlobalPoolChoicesResult, roles=['VIRT_GLOBAL_READ'])
+    async def pool_choices(self):
+        """
+        Pool choices for virtualization purposes.
+        """
+        pools = {'': '[Disabled]'}
+        for p in (await self.middleware.call('zfs.pool.query_imported_fast')).values():
+            if p['name'] in BOOT_POOL_NAME_VALID:
+                continue
+            ds = await self.middleware.call(
+                'pool.dataset.get_instance_quick', p['name'], {'encryption': True},
+            )
+            if not ds['locked']:
+                pools[p['name']] = p['name']
+        return pools
+
+    @private
+    async def internal_interfaces(self):
+        return [INCUS_BRIDGE]
+
+    @private
+    async def get_default_profile(self):
+        result = await incus_call('1.0/profiles/default', 'get')
+        if result.get('status_code') != 200:
+            raise CallError(result.get('error'))
+        return result['metadata']
+
+    @api_method(VirtGlobalGetNetworkArgs, VirtGlobalGetNetworkResult, roles=['VIRT_GLOBAL_READ'])
+    async def get_network(self, name):
+        """
+        Details for the given network.
+        """
+        result = await incus_call(f'1.0/networks/{name}', 'get')
+        if result.get('status_code') != 200:
+            raise CallError(result.get('error'))
+        data = result['metadata']
+        return {
+            'type': data['type'].upper(),
+            'managed': data['managed'],
+            'ipv4_address': data['config']['ipv4.address'],
+            'ipv4_nat': data['config']['ipv4.nat'],
+            'ipv6_address': data['config']['ipv6.address'],
+            'ipv6_nat': data['config']['ipv6.nat'],
+        }
+
+    @private
+    @job()
+    async def setup(self, job):
+        """
+        Sets up incus through their API.
+        Will create necessary storage datasets if required.
+        """
+        try:
+            await self.middleware.call('cache.put', 'VIRT_STATE', 'INITIALIZING')
+            await self._setup_impl(job)
+        except NoPoolConfigured:
+            await self.middleware.call('cache.put', 'VIRT_STATE', 'NO_POOL')
+        except LockedDataset:
+            await self.middleware.call('cache.put', 'VIRT_STATE', 'LOCKED')
+        except Exception:
+            await self.middleware.call('cache.put', 'VIRT_STATE', 'ERROR')
+            raise
+        else:
+            await self.middleware.call('cache.put', 'VIRT_STATE', 'INITIALIZED')
+
+    async def _setup_impl(self, job):
+        config = await self.config()
+
+        if not config['pool']:
+            if await self.middleware.call('service.started', 'incus'):
+                job = await self.middleware.call('virt.global.reset')
+                await job.wait(raise_error=True)
+            await IncusWS().stop()
```

Comment:
I think this start/stop should be delegated to incus service?

We stop incus via `service.stop incus` - expectation should be that everything related has stopped or started?